### PR TITLE
Optimization rule to optimize out nulls for inner joins

### PR DESCRIPTION
**Pull request by jdye64**
Closes: #587 
Closes: #591

### CODE REVIEW

**Review by ayushdg — APPROVED** [review]
Changes lgtm!
A couple of questions around the rule and usage which should be non blocking: 
1. Is there a reason we redefine the rule here instead of directly using: https://github.com/apache/arrow-datafusion/pull/2740. Does that rule not directly work for the planner. 
2. How would we extract this info on the python side? Perhaps modifications to table_scan.rs to expose this info as well?

**Comment by jdye64** [comment]
@ayushdg made the change to use upstream DataFusion version of the optimizer instead of our own. Unfortunately that meant having to refactor some OFFSET/LIMIT code since there were changes to that code in upstream DataFusion. At this point I think everyone should re-review this PR from scratch.

File: dask_sql/physical/rel/logical/limit.py [hunk 1]
```diff
@@ -22,12 +24,55 @@ def convert(self, rel: "LogicalPlan", context: "dask_sql.Context") -> DataContai
...
+        offset = RexConverter.convert(rel, rel.limit().getSkip(), df, context=context)
+
+        # Apply Offset to DataFrame
+        df = self._apply_offset(df, offset) if offset else df
 
-        # If an offset was present it would have already been processed at this point.
-        # Therefore it is always safe to start at 0 when applying the limit
-        df = df.head(limit, npartitions=-1, compute=False)
+        # Limit DataFrame
+        df = df.head(limit, npartitions=-1, compute=False) if limit else df
```
[the remark this answers is not in the document]
Comment (jdye64) [reply dask_sql/physical/rel/logical/limit.py:37]: I don't know? I am not very good at this portion of the codebase to be honest with you. Would you be willing to toss up a PR showing what that would look like?

[the remark this answers is not in the document]
Comment (ayushdg) [reply dask_sql/physical/rel/logical/limit.py:37]: Agree that the original approach might make more sense. Let's open an issue to track this even if we don't prioritize this now.

[the remark this answers is not in the document]
Comment (charlesbluca) [reply dask_sql/physical/rel/logical/limit.py:37]: Opened up #597 to track

[the remark this answers is not in the document]
Comment (jdye64) [reply dask_sql/physical/rel/logical/limit.py:37]: Thanks @charlesbluca !

**Pull request merged**